### ISSUE TITLE
Support `satisfies` operator on functions

### ISSUE THREAD

**Issue by dummdidumm** [microsoft/TypeScript#51556]
# Suggestion

## 🔍 Search Terms

satisfies function

## ✅ Viability Checklist

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

Support the `satisfies` operator to appear after a function

```ts
function foo () { } satisfies Bar
```

This was requested/mentioned in #47920 several times but never followed up on

## 📃 Motivating Example

`satisfies` is super handy to use, and you can already use it for arrow functions. Not having support for it on functions feels unintuitive and forces a certain coding style where people have to use arrow functions when they rather want to use functions.

## 💻 Use Cases

Typing a function while preserving its return type.

**Comment by bradennapier** [comment]
<img width="961" alt="image" src="https://user-images.githubusercontent.com/15365418/202336186-72705a6a-8d51-4f33-83bb-b8c99fba4afc.png">

**Comment by gfx** [comment]
There's already a feature request for the same purpose in 2019: 

* https://github.com/microsoft/TypeScript/issues/34319

**Comment by xamir82** [comment]
Weird that this didn't make it to 4.9, the feature is just incomplete without it.

**Comment by aradalvand** [comment]
This should've honestly been implemented already; such a bummer.

**Comment by PooSham** [comment]
@gfx the difference is that `implements` would require you to define the type of the parameters. But anyone goes, the current situation is quite annoying.

**Comment by ekaradon** [comment]
I was looking for this suggestion. Upvoted it immediately. The first time I had to write some pages with Next, I was just thinking that I would replace lambda by function (for all the benefits that function has over lambda) and thought that satisfies would just work better by not erasing the return type but keeping it type safe nonetheless. Really surprised that an exception for function has been made here. 

 <img src="https://user-images.githubusercontent.com/9439390/222896745-49ee422e-71dc-4626-b1d2-5cfda293351a.png" width="400">


It's such a recurring use case. It would be awesome to get it. Especially because it would be consistent with just all the others use cases where satisfies is valid. Please make it a reality! :smiley:

**Comment by aradalvand** [comment]
It's arguably the most common case where you'd want to use `satisfies` and it's not supported. Brilliant.

**Comment by maslennikov** [comment]
This is what I believe would be the most valuable use-case of `satisfies` for me

**Comment by maslennikov** [comment]
Hey @RyanCavanaugh could you provide any feedback on status of this topic? Could we provide any more feedback to raise its visibility in typescript's backlog?

**Comment by mon-jai** [comment]
Bump

**Comment by xamir82** [comment]
@RyanCavanaugh Any updates?

**Comment by icholy** [comment]
Guys, I'm subscribed to this issue so that I can see updates on its progress. The maintainers are aware of this issue and spamming in here is annoying for everyone involved. Unless you have something meaningful to add the to the conversation, please use the emoji reactions to indicate your support/interest.

**Comment by avshyz-gs** [comment]
bump

**Comment by lrowe** [comment]
This already seems to be supported so long as you add parentheses:
```ts
export default (function f(x: number) { return String(x); } satisfies (x: number) => string);
```

**Comment by francescosalvi** [comment]
> This already seems to be supported so long as you add parentheses:
> 
> ```ts
> export default (function f(x: number) { return String(x); } satisfies (x: number) => string);
> ```

seems to works either way, but perhaps stylistically the closing parenthesis should go right before `satisfies` ?

**Comment by PooSham** [comment]
> This already seems to be supported so long as you add parentheses:
> 
> ```ts
> export default (function f(x: number) { return String(x); } satisfies (x: number) => string);
> ```

Yeah but then you have to type the parameter types again, it would be preferable to use satisfies only for the return type

**Comment by alextbok** [comment]
> > This already seems to be supported so long as you add parentheses:
> > ```ts
> > export default (function f(x: number) { return String(x); } satisfies (x: number) => string);
> > ```
> 
> Yeah but then you have to type the parameter types again, it would be preferable to use satisfies only for the return type

At risk of stating the obvious, if the goal is strictly to declare the parameter type only once in this example (and in general) you can rely on inference from the satisfied type (similar to the return type):

```typescript
export default (function f(x) { return String(x); } satisfies (x: number) => string);
```

**Comment by snarbies** [comment]
Even if there are some slightly awkward ways to manage this with function expressions and arrow functions, it would be nice if there was a solution for function declarations, which seems to be the original ask.

**Comment by wesbos** [comment]
all the solutions have downsides - added parenthesis, using a function expression / arrow function, having to double type the arguments, creating an anonymous function that isn't accessible in the same file (the above example).

We just want to be able to type a functions Params and return types in a single shot and have the values inferred. and satisfies would work?

```ts
type RouteHandler<T extends object> = (body: T) => void;

async function handleRoute({ name }) {
  return new Response.json({ message: `Hello ${name}` });
} satisfies RouteHandler<{ name?: string }>
```

**Comment by PooSham** [comment]
> > > This already seems to be supported so long as you add parentheses:
> > > ```ts
> > > export default (function f(x: number) { return String(x); } satisfies (x: number) => string);
> > > ```
> > 
> > 
> > Yeah but then you have to type the parameter types again, it would be preferable to use satisfies only for the return type
> 
> At risk of stating the obvious, if the goal is strictly to declare the parameter type only once in this example (and in general) you can rely on inference from the satisfied type (similar to the return type):
> 
> ```ts
> export default (function f(x) { return String(x); } satisfies (x: number) => string);
> ```

I wouldn't say that's stating the obvious :) I didn't realize that satisfies on the whole function would make typescript infer the parameter inside the function, but when you mention it it's obvious that typescript needs to do this to ensure type safety. 

I would still prefer the syntax of putting the satisfies keyword for the return type after the parameters, but this is semantically what I'm after. Thank you!

**Comment by louwers** [comment]
If this is implemented, please don't forget about JSDoc.

Example
```ts
export type MyFunc<R extends "a" | "b" | "c" = "a" | "b" | "c"> = () => R;
```

This currently works

```js
/** @satisfies {MyFunc} */
const testB = () => "test";
                 // ^ Type '"test"' is not assignable to type '"a" | "b" | "c"'.ts(2322)
```

This does not:

```js
/** @satisfies {MyFunc} */
function testA() {
  return "test";
}

**Comment by NatoBoram** [comment]
Related:
* https://github.com/eslint/eslint/issues/18842
* https://github.com/typescript-eslint/typescript-eslint/issues/8488

**Comment by RebeccaStevens** [comment]
I think moving the `satisfies` keyword forward would make for easier readability.

e.g.
```ts
function foo satisfies Bar (param1: Baz) { /* ... */ }
```

**Comment by mark-dr** [comment]
@RebeccaStevens I think in that approach the readability would suffer if the type were anything non-trivial. e.g. imagine:

```ts
function foo satisfies NonNullable<SomeInterface>['someMemberName'] (param1: Baz) { /* ... */ }
```

IMO, it creates too much distance between the function name and the parameter list.

**Comment by marcospgp** [comment]
just noticed this isn't possible today - would be really helpful for an API I'm exposing to users if I could just export a function type they could use to both type check *and* type hint their own functions

**Comment by rafa-br34** [comment]
Crazy how it has been 3 years and this is still unimplemented.

**Comment by louwers** [comment]
@rafa-br34 You just sent a notification to 337 (or more) people following this issue without providing any new information. This is bad GitHub etiquette.

**Comment by lolmaus** [comment]
> [@rafa-br34](https://github.com/rafa-br34) You just sent a notification to 337 (or more) people following this issue without providing any new information. This is bad GitHub etiquette.

So did you.

<img width="500" height="500" alt="Image" src="https://github.com/user-attachments/assets/1591c533-6cc1-4fdc-a7f2-5f81cff41d61" />

TypeScript is open source but it's **not a non-profit product**. Microsoft are earning many, many millions from it in the long run, otherwise they wouldn't be making it in the first place.

The argument "it's open source, fix it yourself or shut the funk up" just does not apply. We, individual developers, nether have capacity nor expertise for it. The corporation have both and choose to ignore it.

Being loud about our needs is literally the only way for us to make any difference.
